### PR TITLE
Ensure names generated by auto-complete are unique

### PR DESCRIPTION
**Pull request by darzu**
Fixes: https://github.com/microsoft/pxt-minecraft/issues/1479

- In SyntaxInfo, keep a seperate list of global symbols for all types of requests so that we can ensure uniqueness for member completion or identifier completion
- Update py2ts to return these
- Update getCompletions and getSnippet to account for existing symbols
- remove dead code, improve comments, rename a few variables

### CODE REVIEW

File: pxtpy/converter.ts [hunk 1]
```diff
@@ -2433,6 +2434,25 @@ namespace pxt.py {
             if (!syntaxInfo.symbols)
                 syntaxInfo.symbols = []
 
+            // always return global symbols because we might need to check for
+            // name collisions downstream
+            {
```
Comment (riknoll) [inline pxtpy/converter.ts]: why is this in a block?

File: pxtcompiler/emitter/service.ts [hunk 2]
```diff
@@ -833,10 +804,29 @@ namespace ts.pxtc.service {
...
+        success: boolean,
+        outfiles: { [key: string]: string },
+        syntaxInfo?: pxtc.SyntaxInfo
+    }
+    function updatePySyntaxInfo(opts: CompileOptions) {
+        // TODO: We call py2ts here directly to get the python syntax info.
+        // We likely already have the syntax info from a previous compile,
+        // we should do better about cache the latest results and serving those
+        // here. Also, we shouldn't mutate "opts" (the input).
+        let res = (pxt as any).py.py2ts(opts) as Py2TsRes
```
Comment (riknoll) [inline pxtcompiler/emitter/service.ts]: Why the `any` cast here?

File: pxtcompiler/emitter/service.ts [hunk 3]
```diff
@@ -833,10 +804,29 @@ namespace ts.pxtc.service {
         }
     }
 
+    type Py2TsRes = {
```
Comment (darzu) [inline pxtcompiler/emitter/service.ts]: swaping:
image

File: pxtcompiler/emitter/service.ts [hunk 4]
```diff
@@ -1254,22 +1248,31 @@ namespace ts.pxtc.service {
 . . . . .
 \``;
 
-    export function getSnippet(apis: ApisInfo, runtimeOps: pxt.RuntimeOptions, fn: SymbolInfo, n: ts.FunctionLikeDeclaration, python?: boolean): string {
+    export function getSnippet(apis: ApisInfo, takenNames: pxt.Map<SymbolInfo>, runtimeOps: pxt.RuntimeOptions, fn: SymbolInfo, decl: ts.FunctionLikeDeclaration, python?: boolean): string {
```
Comment (riknoll) [inline pxtcompiler/emitter/service.ts]: We should consider refactoring this at some point; too many arguments

Comment (darzu) [reply pxtcompiler/emitter/service.ts]: I definitely agree. I spent a good hour this morning trying to unify lastApisInfo and this syntax info stuff and I had to bail, it was too tangle.

File: pxtcompiler/emitter/service.ts (hunk 2 above)
Comment (darzu) [reply pxtcompiler/emitter/service.ts]: I forget the justification, but Michal tried for a while to get references to the python code to work right with TSC and I believe I gave up. I think I tried at one point too. If you have ideas, I'm all ears.

File: pxtpy/converter.ts (hunk 1 above)
Comment (darzu) [reply pxtpy/converter.ts]: Used to have an "if",  i'll remove the extra block

File: pxtcompiler/emitter/service.ts (hunk 4 above)
Comment (darzu) [reply pxtcompiler/emitter/service.ts]: I left a number of new TODO comments detailing some of the things I think are broken / inflexible with our current approach

**Pull request merged**